### PR TITLE
Add InfoTree pydantic models

### PR DESCRIPTION
**Pull request by sorgfresser**

### CODE REVIEW

**Comment by sorgfresser** [comment]
Waiting for https://github.com/leanprover-community/repl/pull/89

**Comment by sorgfresser** [comment]
Rebasing should (ideally) have fixed it. I locally don't get any errors.

**Comment by augustepoiroux** [comment]
Great thanks! Running tests
I noticed that there is `CommmandInfo` in the code. I believe this typo has been fixed in the REPL since then (`CommandInfo`)

**Comment by sorgfresser** [comment]
Strange that this didn't raise any issues for me... Taking a look, sorry!

File: src/lean_interact/interface.py [hunk 1]
```diff
@@ -326,7 +341,7 @@ def commands(self) -> Generator[Self, None, None]:
             Yields the command nodes of the InfoTree.
         """
         for tree in self.dfs_walk():
-            if tree.kind != "CommmandInfo":
+            if tree.kind != "CommandInfo":
```
Comment (sorgfresser) [inline src/lean_interact/interface.py:344]: Nice catch!

**Comment by augustepoiroux** [comment]
Tests are passing +1
However, my type checker was complaining about some edge cases not handled properly. I fixed these issues.
It would probably be nice to add some tests for the InfoTree methods you introduced ;)

**Comment by sorgfresser** [comment]
Adding tests, one sec

**Comment by sorgfresser** [comment]
> Oh sorry, most of these "issues" were false positives

Yeah, I think the `Commmand` one was pretty bad, the rest is not really necessary? self.node can't be none...
Thanks for catching this one though!

File: src/lean_interact/interface.py [hunk 2]
```diff
@@ -198,6 +200,248 @@ def message_intersects_code(msg: Message | Sorry, start_pos: Pos | None, end_pos
...
+        elaborator: Optionally, the elaborator used.
+    """
+
+    is_binder: bool = Field(alias="isBinder")
+    expr: str
+    expected_type: str | None = Field(default=None, alias="expectedType")
+    elaborator: str | None
+
+
+Node = TacticNode | CommandNode | TermNode | None
```
Comment (augustepoiroux) [inline src/lean_interact/interface.py:283]: `self.node` can be `None` according to its typing annotation (in Lean, node is an option as well).
Yeah, otherwise, most of the checks I added are useless.
I guess something better to do would be to update the return type of the `commands` method to tell the user, and the type checker, it is a generator of `CommandNode`.

Comment (augustepoiroux) [reply src/lean_interact/interface.py:283]: Oh, but this is not trivial. It is returning `InfoTree`s with nodes that are `CommandNode`...

**Comment by sorgfresser** [comment]
Added the tests now!

File: src/lean_interact/interface.py (hunk 2 above)
Comment (augustepoiroux) [reply src/lean_interact/interface.py:283]: Yeah I agree, let's merge like this

File: tests/test_server.py [hunk 3]
```diff
@@ -549,17 +549,21 @@ def test_infotree_theorems(self):
         self.assertIsInstance(result, CommandResponse)
         assert isinstance(result, CommandResponse)
         self.assertIsNotNone(result.infotree)
+        assert result.infotree is not None
```
[the remark this answers is not in the document]
Comment (augustepoiroux) [reply tests/test_server.py:552]: Yes, just like the assert with `CommandResponse` above. I agree this is redundant and ugly ^^ But the rest of the test script is like this as well. Type checkers do not understand the `self.assert...` methods. Since we don't care about performance in the tests, I prefer making type checkers happy with proper asserts instead of having `# type: ignore` everywhere.

[the remark this answers is not in the document]
Comment (sorgfresser) [reply tests/test_server.py:552]: Fair point, fine by me! :)
